experimental/packages/otlp-transformer/src/trace/index.ts

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import type { InstrumentationLibrary } from '@opentelemetry/core';
+import type { Resource } from '@opentelemetry/resources';
+import type { ReadableSpan } from '@opentelemetry/sdk-trace-base';
+import { toAttributes } from '../common/internal';
+import { toSpan } from './internal';
+import { IExportTraceServiceRequest } from './types';
+
+export function createExportTraceServiceRequest(spans: ReadableSpan[]): IExportTraceServiceRequest | null {
+  if (spans.length === 0) {
+    return null;
+  }
+
+  return {
+    resourceSpans: groupByResource(spans).map(({ resource, resourceSpans, schemaUrl: resourceSchemaUrl }) => ({
+      resource: {
+        attributes: toAttributes(resource.attributes),
+        droppedAttributesCount: 0,
+      },
+      instrumentationLibrarySpans: resourceSpans.map(({ instrumentationLibrary, instrumentationLibrarySpans, schemaUrl: librarySchemaUrl }) => ({
+        instrumentationLibrary: {
+          name: instrumentationLibrary.name,
+          version: instrumentationLibrary.version,
+        },
+        spans: instrumentationLibrarySpans.map(toSpan),
+        schemaUrl: librarySchemaUrl,
+      })),
+      schemaUrl: resourceSchemaUrl,
+    }))
+  };
+}
+
+type ResourceSpans = {
+  resource: Resource,
+  resourceSpans: InstrumentationLibrarySpans[],
+  schemaUrl?: string,
+};
+
+type InstrumentationLibrarySpans = {
+  instrumentationLibrary: InstrumentationLibrary,
+  instrumentationLibrarySpans: ReadableSpan[],
+  schemaUrl?: string,
+};
+
+function groupByResource(SpanRecords: ReadableSpan[]): ResourceSpans[] {
+  const resourceMap: Map<Resource, Map<string, ReadableSpan[]>> = new Map();
+
+  for (const record of SpanRecords) {
+    let ilmMap = resourceMap.get(record.resource);
+
+    if (!ilmMap) {
+      ilmMap = new Map();
+      resourceMap.set(record.resource, ilmMap);
+    }
+
+    const instrumentationLibraryKey = JSON.stringify([record.instrumentationLibrary.name, record.instrumentationLibrary.version, record.instrumentationLibrary.schemaUrl]);
```

Comment:
we already have this logic in `BasicTracerProvider` and it is [implemented like this](https://github.com/open-telemetry/opentelemetry-js/blob/c6dab2a6e721208c7ef08c85d8a80c0870e8a8b3/packages/opentelemetry-sdk-trace-base/src/BasicTracerProvider.ts#L95):
```js
    const key = `${name}@${version || ''}:${options?.schemaUrl || ''}`;
```

Consider if we want to consolidate the logic here or there.

Reply:
consolidating logic would require one depending on the other. maybe it should be in core?

Reply:
In this case it is important that the key is JSON because it is parsed later. I will try to make it not serialize and parse as it could easily be accidentally broken in the future.

Reply:
Yeah I have seen it later. I'm in favor of not stringifying and parsing as I wrote in a different comment if it's not too complicated and makes sense to you.